### PR TITLE
many: add Is{Core,Classic}Boot() to DeviceContext

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Is{Core,Classic}Boot() will be used to find if we are in a Ubuntu Core
device and if we are in a classic with modes (there is a modeenv file)
device respectively. That is determined by the information found in
the model assertion.

### CODE REVIEW

File: boot/boottest/device.go [hunk 1]
```diff
@@ -46,30 +47,52 @@ func MockDevice(s string) snap.Device {
...
-// MockUC20Device implements boot.Device and returns true for HasModeenv.
-// Arguments are mode (empty means "run"), and model.
-// If model is nil a default model is used (same as MakeMockUC20Model).
-func MockUC20Device(mode string, model *asserts.Model) snap.Device {
+// MockDeviceWithModes implements boot.Device and returns true for
+// HasModeenv.  Arguments are mode (empty means "run"), model, and a
+// boolean specifying if this is a classic with modes or a UC device.
+// If model is nil a default model is used (MakeMockUC20Model or
+// MakeMockClassicWithModesModel is called).
+func MockDeviceWithModes(mode string, model *asserts.Model, isClassic bool) snap.Device {
```
Comment (pedronis) [inline boot/boottest/device.go:62]: mmh, it seems this should be a not exported helper

File: boot/boottest/device.go [hunk 2]
```diff
@@ -46,30 +47,52 @@ func MockDevice(s string) snap.Device {
...
+		hasModes:  true,
+		isClassic: isClassic,
+		model:     model,
 	}
 }
 
+// MockUC20Device mocks a UC with modes device.
+// Arguments are mode (empty means "run"), and model.
+func MockUC20Device(mode string, model *asserts.Model) snap.Device {
+	isClassic := false
```
Comment (pedronis) [inline boot/boottest/device.go:85]: this shoulld panic if model is passed and is classic?

File: boot/boottest/device.go [hunk 3]
```diff
@@ -46,30 +47,52 @@ func MockDevice(s string) snap.Device {
...
+	return MockDeviceWithModes(mode, model, isClassic)
+}
+
+// MockClassicBootDevice mocks a classic with modes device.
+// Arguments are mode (empty means "run"), and model.
+func MockClassicBootDevice(mode string, model *asserts.Model) snap.Device {
+	isClassic := true
```
Comment (pedronis) [inline boot/boottest/device.go:92]: this instead should panic if model is passed in and is not classic?

File: overlord/devicestate/devicectx.go [hunk 4]
```diff
@@ -104,6 +104,17 @@ func (dc groundDeviceContext) HasModeenv() bool {
 	return dc.model.Grade() != asserts.ModelGradeUnset
 }
 
+// IsCoreBoot is true when there are modes, or when there are not but
+// we are not in classic (UC16/18 case)
+func (d *groundDeviceContext) IsCoreBoot() bool {
```
Comment (pedronis) [inline overlord/devicestate/devicectx.go:109]: we don't have it but maybe it is time to start a devicectx_test.go ?

**Review by pedronis — COMMENTED** [review]
thanks, did a pass

File: overlord/devicestate/devicectx.go [hunk 5]
```diff
@@ -104,6 +104,17 @@ func (dc groundDeviceContext) HasModeenv() bool {
...
+	return d.HasModeenv() || !d.Classic()
+}
+
+// IsClassicBoot is true for classic systems with classic initramfs
+func (d *groundDeviceContext) IsClassicBoot() bool {
+	return !d.IsCoreBoot()
```
Comment (mardy) [inline overlord/devicestate/devicectx.go:115]: This expands to
```go
!(d.HasModeenv() || !d.Classic())
```
which is:
```go
(!d.HasModeenv() && d.Classic())
```
that is a classic system with no run mode set. I'm a bit ignorant of the boot process details (I guess we don't use run modes on classic? Or do we have the "install" mode there too?), so it's likely that this is correct, but I just wanted to double check with you. I wonder if the function docstring could be expanded a bit to clarify this (it mentions the initramfs, but no run mode).

**Review by mardy — APPROVED** [review]
LGTM, with a question.

File: overlord/devicestate/devicectx.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicectx.go:115]: This PR is part of the changes for starting to use the UC initramfs on classic systems. So, we did not have modes for classic in the past, but now we will start to distinguish betwee classic with and without modes. I'll clarify a bit more in the docstring.

File: boot/boottest/device.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply boot/boottest/device.go:98]: Done

File: overlord/devicestate/devicectx.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicectx.go:109]: Added now

File: boot/boottest/device.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply boot/boottest/device.go:88]: Done

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, this is ready for another round now.

File: boot/boottest/device.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply boot/boottest/device.go]: Right, I've unexported now

File: overlord/devicestate/devicectx_test.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		"retrieved ground context is not intended to drive store operations")
+	c.Check(devCtx.ForRemodeling(), Equals, false)
+	c.Check(devCtx.SystemMode(), Equals, "run")
+}
+
+func (s *deviceCtxSuite) TestGroundDeviceContext(c *C) {
+	var devCtx snapstate.DeviceContext
+
+	// Classic with classic initramfs
+	classModel := assertstest.FakeAssertion(map[string]interface{}{
```
Comment (pedronis) [inline overlord/devicestate/devicectx_test.go:58]: I wouldn't abbreviate classic to class, as class is a proper word as well with  a different meaning. I think is best to call this classicModel, same below

File: overlord/devicestate/devicectx_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	c.Check(devCtx.Base(), Equals, "core20")
+	c.Check(devCtx.Gadget(), Equals, "pc")
+	c.Check(devCtx.HasModeenv(), Equals, true)
+	c.Check(devCtx.IsCoreBoot(), Equals, true)
+	c.Check(devCtx.IsClassicBoot(), Equals, false)
+	c.Check(devCtx.Model(), DeepEquals, ucModel)
+	testGroundDeviceContextCommonChecks(c, devCtx)
+
+	// Classic with modes
+	classWithModes := boottest.MakeMockClassicWithModesModel()
```
Comment (pedronis) [inline overlord/devicestate/devicectx_test.go:109]: same issue with the abbreviation

**Review by pedronis — APPROVED** [review]
lgtm, some small remarks

**Comment by alfonsosanchezbeato** [comment]
@pedronis thanks for the reviews, I've addressed now your latest comments.

File: overlord/devicestate/devicectx.go (hunk 5 above)
Comment (mardy) [reply overlord/devicestate/devicectx.go:118]: It's clear now, thanks!

**Review by mardy — APPROVED** [review]

**Pull request merged**